File: backend_nodejs/src/application_service/usecase/markAsCompletedTodoUsecase.ts
import type { Todo } from "@/domain/model/todo";
import type { TodoRepository } from "@/domain/repository/todoRepository";
import type { TransactionService } from "../service/transactionService";

export type MarkAsCompletedTodoRequest = {
  id: string;
};

export type MarkAsCompletedTodoResponse = {
  todo: Todo;
};

export interface MarkAsCompletedTodoUsecase {
  execute(
    request: MarkAsCompletedTodoRequest,
  ): Promise<MarkAsCompletedTodoResponse>;
}

export class MarkAsCompletedTodoUsecaseImpl
  implements MarkAsCompletedTodoUsecase
{
  constructor(
    private todoRepository: TodoRepository,
    private transactionService: TransactionService,
  ) {}

  async execute(
    request: MarkAsCompletedTodoRequest,
  ): Promise<MarkAsCompletedTodoResponse> {
    const todo = await this.transactionService.run(async () => {
      const todo = await this.todoRepository.findById(request.id);
      todo.markAsCompleted();
      await this.todoRepository.save(todo);
      return todo;
    });
    return { todo };
  }
}

File: backend_python/src/todo_api/application_service/usecase/mark_as_completed_todo_usecase.py
from abc import ABC, abstractmethod
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from todo_api.application_service.service.transaction_service import TransactionService
from todo_api.domain.model.todo import Todo
from todo_api.domain.repository.todo_repository import TodoRepository


class MarkAsCompletedTodoUsecaseInput(BaseModel):
    id: UUID


class MarkAsCompletedTodoUsecaseOutput(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    todo: Todo


class MarkAsCompletedTodoUsecase(ABC):
    @abstractmethod
    def execute(
        self, input_dto: MarkAsCompletedTodoUsecaseInput
    ) -> MarkAsCompletedTodoUsecaseOutput:
        pass


class MarkAsCompletedTodoUsecaseImpl(MarkAsCompletedTodoUsecase):
    def __init__(
        self, todo_repository: TodoRepository, transaction_service: TransactionService
    ) -> None:
        self.todo_repository = todo_repository
        self.transaction_service = transaction_service

    def execute(
        self, input_dto: MarkAsCompletedTodoUsecaseInput
    ) -> MarkAsCompletedTodoUsecaseOutput:
        def func() -> MarkAsCompletedTodoUsecaseOutput:
            todo = self.todo_repository.find_by_id(input_dto.id)
            todo.mark_as_completed()
            self.todo_repository.save(todo)
            return MarkAsCompletedTodoUsecaseOutput(todo=todo)

        return self.transaction_service.Run(func)
